Add IPv6 support for Matomo Installation

nathangavin opened the pull request:
Adding IPv6 support for Matomo Installation

### Description:

Resolves issue #22677.  Current bug is that users cannot provide an IPv6 address when installing Matomo. This fix addresses it by expanding the logic around processing the user input for database server, and if IPv6 is detected then its formatted correctly.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

nathangavin committed:
extended database server processing during installation to account for IPv6 addresses

nathangavin commented:
Milestone will be set to 5.2.1 when it is created.

caddoo commented:
@nathangavin Thanks for providing this PR.

I have a few suggestions to consider that might improve the readability and maintainability of this code. Let me know your thoughts:

**Clarity and Readability:** At first glance, it's not immediately clear what the entire block of code is doing. To understand it fully, I would need to read through each part in detail. This could be challenging for someone revisiting this code later. Do you think we could benefit from breaking it into smaller, more descriptive functions or methods with names that convey the intent of each part?

**Function/Method Extraction:** Extracting specific logic into smaller functions or even a dedicated class might make the overall flow easier to follow. This could also help reduce cognitive load for future developers working on this.

**Testing Considerations:** Since this seems to be a crucial part of the application—potentially affecting users' ability to install—have you thought about adding tests for this? Unit tests or integration tests could help ensure the reliability of this code and catch any edge cases. If there are existing tests for related functionality, it might be worth reviewing or extending them as well.


Let me know if you’d like to discuss this further, I’m happy to collaborate on brainstorming ways to refine it. Great work so far!

nathangavin committed:
Move host & port extraction into its own function

nathangavin committed:
Extract logic for each host type into functions

nathangavin committed:
Move Host & Port extraction logic to own class

Migrated the host and port extraction logic to its own class to
facilitate easier testing, and to reduce the amount of functionality
contained withing the FormDatabaseSetup class.

nathangavin committed:
Implement suggested changes from PHPCS

nathangavin committed:
Add handling for extraction failure

Modified the extract function so that it can return a nullable. This is
in place to be returned when no values can be successfully extracted
from the provided host.

nathangavin committed:
Implement test suite for HostPortExtractor

nathangavin committed:
Makes code 7.2 compliant, fixes PHPCS issues

nathangavin committed:
Fix another 7.2 issue, more PHPCS

nathangavin committed:
Fix PHPCS issue

nathangavin committed:
Fix another PHPCS issue with multiline

michalkleiner commented on plugins/Installation/HostPortExtractor.php:
```diff
@@ -0,0 +1,182 @@
+        $components = explode(']', $dbHost);
+        if (count($components) > 1) {
+            // check if the text is a valid port e.g. ':3000'
+            preg_match('/^(:\d+)$/', $components[1], $portMatches);
+            if (count($portMatches) > 0) {
+                // only save the port number, not the colon
+                $port = substr($portMatches[1], 1);
+            }
```
we don't need to capture the `:`, just the number
```suggestion
            preg_match('/^:(\d+)$/', $components[1], $portMatches);
            if (count($portMatches) > 0) {
                // only save the port number, not the colon
                $port = $portMatches[1];
            }
```

nathangavin replied:
The regex is checking to make sure the text retrieved from after the closing square bracket correctly matches the port format. so with "]:3000", it would find ":3000" which is valid, but "]3000" is not a valid port format. 

Though we should probably return null if content is found after the square brackets but it doesn't match the port format, as this indicates an invalid host has been provided.

michalkleiner replied:
I suggested changing the capturing group, otherwise the regex matches the same thing as before.

nathangavin replied:
Oh I see, that makes sense I will make that change.

nathangavin replied:
I have addressed this issue.

michalkleiner commented on plugins/Installation/HostPortExtractor.php:
```diff
@@ -0,0 +1,182 @@
+     */
+    private static function isUnixSocket(string $dbHost): bool
+    {
+        return (strpos($dbHost, '/') !== false);
```
Is the check meant to be that the string starts with a slash? Or that it has a slash anywhere?
```suggestion
        return (strpos($dbHost, '/') === 0);
```

nathangavin replied:
This is the original logic so I've changed it as little as possible, however yes the string should be required to start with a slash, to match "/path/sock.n" format.

michalkleiner replied:
I see re: the move of the code without changing the logic.
I think that's a valid small improvement we can do here unless sockets can be defined in some other way where there's no leading `/` and it's prefixed with something else before that. Let's keep it as is.

michalkleiner commented on plugins/Installation/HostPortExtractor.php:
```diff
@@ -0,0 +1,182 @@
+            throw new InvalidArgumentException('$dbHost must be a valid Unix Socket');
+        }
+
+        $portIndex = strpos($dbHost, '/');
+            // unix_socket=/path/sock.n
+        $port = substr($dbHost, $portIndex);
```
Not 100% sure here, are we considering the socket being the port here? Should the variables be named differently or perhaps adjust the comment explaining the specificity of this?

nathangavin replied:
The DB connector requires unix sockets to be in the 'port' field. I will add a comment explaining this.

nathangavin replied:
I have made this change.

michalkleiner commented on plugins/Installation/HostPortExtractor.php:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    /**
+     * Determines if the providedhost is a standard web address or IP with a port
```
```suggestion
     * Determines if the provided host is a standard web address or IP with a port
```

nathangavin replied:
Fixed.

michalkleiner commented on plugins/Installation/HostPortExtractor.php:
```diff
@@ -0,0 +1,182 @@
+            throw new InvalidArgumentException('$dbHost must be a valid address');
+        }
+
+        $hostAndPort = explode(':', $dbHost);
+        $host = $hostAndPort[0];
+        $port = $hostAndPort[1];
```
```suggestion
        [$host, $port] = explode(':', $dbHost);
```

nathangavin replied:
Implemented

michalkleiner commented on plugins/Installation/tests/Unit/HostPortExtractorTest.php:
Good coverage of the various inputs, I'd probably also add a short IPv6 with a port to complete the set.

Another improvement that can make it easier to add new test cases is to use a data provider and a single test method since we always test the same thing - get an input, extract and compare the output.

Have a look at e.g. `tests/PHPUnit/Unit/VersionTest.php` and check how `@dataProvider getLowerVersionCompares` annotation works together with `testVersionCompares` test method.

michalkleiner replied:
So the data provider can return an array of this structure
```
[
    ['input1', ['host1', 'port1']],
    ['invalid input', null],
    ['ipv6:4000', ['ipv6', '4000']],
]
```
and then the single test method can take two params, dbHost and hostAndPort, and make the assertions on those. Or have two data providers, one for valid and one for invalid inputs, and have two test methods, one with the host and port params, and one with just the input as the assertion would be to check null which we don't need to pass in.

nathangavin replied:
I have implemented the data provider approach and refactored the testing accordingly.

michalkleiner commented on plugins/Installation/HostPortExtractor.php:
```diff
@@ -0,0 +1,182 @@
+    private static function isIPv6(string $dbHost): bool
+    {
+        // IPv6 addresses must contain '[' & ']' to be considered valid
+        if (strpos($dbHost, '[') !== false && strpos($dbHost, ']') !== false) {
```
Perhaps we can just directly use regex to check the string starts and ends with `[` and `]` and extract the text at the same time?

With the current implementation this string would be considered valid `abc [::] def`.

michalkleiner replied:
You can add strings like that into the test and then adjust the implementation until it passes.

nathangavin replied:
I have removed the first check for square brackets and improved the following regex to do the necessary checks. I have added a test for this scenario.

michalkleiner commented:
Nice job @nathangavin, I really like the extractor class approach. It makes the code tidier and more testable. I've left a few comments to consider, but nothing major. Well done!

nathangavin committed:
Streamline handling of explosion

nathangavin committed:
Explain how unix sockets need to be handled

nathangavin committed:
Update tests to use data providers

nathangavin committed:
Add extra edge cases

nathangavin committed:
Fix issues highlighted by testing

nathangavin committed:
Fix PHPCS issues

michalkleiner reviewed (approved):
The changes look good to me thought I haven't tested it locally as I don't have an IPv6 setup running.
